File: Programmers/Sort/42746.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

#define DEBUG

class NumInfo {
public:
    int index;
    int number;
    int original;
};

string solution(vector<int> numbers) {
    string answer = "";

    // index, number value
    vector<NumInfo> numIdxs;
    int idx = 0;
    for (auto number : numbers) {
        int original = number;
        int lastNum = number % 10;

        if (number < 10) {
            number = number * 1000 + lastNum * 111;
        }
        else if (10 <= number && number < 100) {
            number = number * 100 + lastNum * 11;
        }
        else if (100 <= number && number < 1000) {
            number = number * 10 + lastNum;
        }

        numIdxs.push_back(NumInfo { idx, number, original });
    }

    sort(numIdxs.begin(), numIdxs.end(), [] (NumInfo ls, NumInfo rs) -> bool {
        if (ls.number == rs.number) {
            return ls.original < rs.original;
        }
        return ls.number > rs.number;
    });

    #ifdef DEBUG
    for (auto numIdx : numIdxs) {
        cout << numIdx.second << endl;
    }
    #endif

    for (auto numIdx : numIdxs) {
        answer += to_string(numbers[numIdx.first]);
    }

    return answer;
}

File: Programmers/Level2/42746.js
function solution(numbers) {
  if (numbers.length === 0) return "";

  const newNumbers = numbers.map((number) => {
    if (number < 10) {
      return {
        number,
        comp: 1111 * number,
      };
    } else if (number >= 10 && number < 100) {
      return {
        number,
        comp: number * 101,
      };
    } else if (number >= 100 && number < 1000) {
      return {
        number,
        comp: number * 10 + number / 100,
      };
    } else
      return {
        number,
        comp: number,
      };
  });

  const nums = newNumbers
    .sort((a, b) => b.comp - a.comp)
    .map((number) => number.number);
  if (nums[0] === 0) return "0";
  else return nums.join("");
}
